src/cart/crt_context.c

```diff
@@ -616,45 +606,73 @@ crt_ctx_epi_abort(d_list_t *rlink, void *arg)
 	return rc;
 }
 
-/* Free the last reference of ctx. */
-void
-crt_context_free(struct crt_context *ctx)
+/* See crt_rank_abort. */
+static pthread_rwlock_t crt_context_destroy_lock = PTHREAD_RWLOCK_INITIALIZER;
+
+static int
+crt_context_epis_append(d_list_t *rlink, void *arg)
 {
-	int	provider = ctx->cc_hg_ctx.chc_provider;
-	int	rc;
+	struct d_vec_pointers	*epis = arg;
+	struct crt_ep_inflight	*epi = epi_link2ptr(rlink);
+	int			 rc;
 
-	rc = d_hash_table_destroy_inplace(&ctx->cc_epi_table, true /* force */);
-	if (rc)
-		D_ERROR("destroy context (idx %d), d_hash_table_destroy_inplace failed: "DF_RC"\n",
-			ctx->cc_idx, DP_RC(rc));
+	d_hash_rec_addref(&epi->epi_ctx->cc_epi_table, rlink);
+	rc = d_vec_pointers_append(epis, epi);
+	if (rc != 0)
+		d_hash_rec_decref(&epi->epi_ctx->cc_epi_table, rlink);
+	return rc;
+}
 
-	d_binheap_destroy_inplace(&ctx->cc_bh_timeout);
+static int
+crt_context_abort(struct crt_context *ctx, bool force)
+{
+	struct d_vec_pointers	epis;
+	int			flags;
+	int			i;
+	int			rc;
 
-	rc = crt_hg_ctx_fini(&ctx->cc_hg_ctx);
-	if (rc)
-		D_ERROR("crt_hg_ctx_fini failed() rc: " DF_RC "\n", DP_RC(rc));
+	rc = d_vec_pointers_init(&epis, 16 /* cap */);
+	if (rc != 0)
+		D_GOTO(out, rc);
 
-	D_RWLOCK_WRLOCK(&crt_gdata.cg_rwlock);
+	D_MUTEX_LOCK(&ctx->cc_mutex);
+	rc = d_hash_table_traverse(&ctx->cc_epi_table, crt_context_epis_append, &epis);
+	D_MUTEX_UNLOCK(&ctx->cc_mutex);
+	if (rc != 0)
+		D_GOTO(out_epis, rc);
 
-	crt_provider_put_ctx_idx(ctx->cc_primary, provider, ctx->cc_idx);
-	d_list_del(&ctx->cc_link);
+	flags = force ? (CRT_EPI_ABORT_FORCE | CRT_EPI_ABORT_WAIT) : 0;
 
-	D_RWLOCK_UNLOCK(&crt_gdata.cg_rwlock);
+	for (i = 0; i < epis.p_len; i++) {
+		struct crt_ep_inflight *epi = epis.p_buf[i];
 
-	D_MUTEX_DESTROY(&ctx->cc_mutex);
-	D_DEBUG(DB_TRACE, "destroyed context (idx %d)\n", ctx->cc_idx);
-	D_FREE(ctx);
+		rc = crt_ctx_epi_abort(epi, flags);
+		if (rc != 0)
+			break;
+	}
+
+out_epis:
+	for (i = 0; i < epis.p_len; i++) {
+		struct crt_ep_inflight *epi = epis.p_buf[i];
+
+		d_hash_rec_decref(&ctx->cc_epi_table, &epi->epi_link);
```

Comment:
should this be wrapped in D_MUTEX_LOCK/UNLOCK on ctx->cc_mutex?

Reply:
Indeed. Fixed.